rule/expr.go

```diff
@@ -29,13 +36,7 @@ type Params interface {
 	GetFloat64(key string) (float64, error)
 	Keys() []string
 	EncodeValue(key string) (string, error)
-}
-
-// Param is an expression used to select a parameter passed during evaluation and return its corresponding value.
-type Param struct {
-	Kind string `json:"kind"`
-	Type string `json:"type"`
-	Name string `json:"name"`
+	AddParam(key string, value interface{}) (Params, error)
```

Comment:
I was thinking maybe we should consider parameters and variables as a stack (only in the `rule` package).
`context.Context` elegantly uses embedding to implement a stack without copy, maybe we could do something like that?

Currently, all the `Params` methods are for reading values; for writes we could do something like that:

```go
type paramStack struct {
  Params Params

  key string
  value interface{}
}

func (p paramStack) GetFloat64(key string) (float64, error) {
  if p.key == key {
    return p.value.(float64), nil
  }

  return p.Params.GetFloat(key)
}

...

func addParam(params Params, key string, value interface{}) Params {
  return &paramStack{
    Params: params,
    Key: key,
    Value: value,
  }
}
```

That way, we have the following benefits:
- we don't have to implement an AddParam method to the `Params` interface
- we have the certitude params are not mutated during evaluation, since we are not depending on the their underlying `AddParam` implementation but ours
- we avoid copy
- upper scopes don't even know new variables have been created in lower scopes

WDYT?